packages/react-aria-components/src/GridList.tsx

```diff
@@ -99,37 +254,133 @@ function GridListItem({item}) {
+      allowsDragging: !!dragState,
+      isDragging,
+      isDropTarget: dropIndicator?.isDropTarget
+    }
+  });
+
+  let renderDropIndicator = dragAndDropHooks?.renderDropIndicator || (target => <DropIndicator target={target} />);
+  let dragButtonRef = useRef(null);
+  useEffect(() => {
+    if (dragState && !dragButtonRef.current) {
+      console.warn('Draggable items in a GridList must contain a <Button slot="drag"> element so that keyboard and screen reader users can drag them.');
+    }
+  // eslint-disable-next-line
+  }, []);
+
+  return (
+    <>
+      {dragAndDropHooks?.useDropIndicator &&
+        renderDropIndicator({type: 'item', key: item.key, dropPosition: 'before'})
+      }
+      {dropIndicator && !dropIndicator.isHidden &&
+        <div role="row">
+          <div role="gridcell">
+            <div role="button" {...visuallyHiddenProps} {...dropIndicator?.dropIndicatorProps} ref={dropIndicatorRef} />
+          </div>
+        </div>
+      }
+      <div
+        {...mergeProps(rowProps, focusProps, hoverProps, draggableItem?.dragProps)}
+        {...renderProps}
+        ref={ref}
+        data-hovered={isHovered || undefined}
+        data-focused={states.isFocused || undefined}
+        data-focus-visible={isFocusVisible || undefined}
+        data-pressed={states.isPressed || undefined}
+        data-dragging={isDragging || undefined}
+        data-drop-target={dropIndicator?.isDropTarget || undefined}>
+        <div {...gridCellProps}>
+          <Provider
+            values={[
+              [CheckboxContext, checkboxProps],
+              [ButtonContext, {
+                slots: {
+                  [defaultSlot]: {},
+                  drag: {
+                    ...draggableItem?.dragButtonProps,
+                    ref: dragButtonRef,
+                    style: {
+                      pointerEvents: 'none'
+                    }
+                  }
+                }
+              }],
+              [TextContext, {
+                slots: {
+                  description: descriptionProps
+                }
+              }]
+            ]}>
+            {renderProps.children}
+          </Provider>
+        </div>
+      </div>
+      {dragAndDropHooks?.useDropIndicator && state.collection.getKeyAfter(item.key) == null &&
+        renderDropIndicator({type: 'item', key: item.key, dropPosition: 'after'})
+      }
+    </>
+  );
+}
+
+function GridListDropIndicator(props: DropIndicatorProps, ref: ForwardedRef<HTMLElement>) {
+  ref = useObjectRef(ref);
+  let {dragAndDropHooks, dropState} = useContext(InternalGridListContext);
+  let buttonRef = useRef(null);
+  let {dropIndicatorProps, isHidden, isDropTarget} = dragAndDropHooks.useDropIndicator(
+    props,
+    dropState,
+    buttonRef
+  );
+
+  let {visuallyHiddenProps} = useVisuallyHidden();
+
+  if (isHidden) {
+    return null;
+  }
+
+  // eslint-disable-next-line react-hooks/rules-of-hooks
+  let renderProps = useRenderProps({
```

Comment:
useRenderProps isn't that expensive, can we just move this before the isHidden check?

Reply:
I wrote this a long time ago but I think the reason was that users can pass functions to the `className`, `style`, and `children` props, and we'd still be calling them even when hidden and actually rendering nothing so might be weird?

Reply:
ah, i think this pattern was taken from elsewhere? I thought it was tailwind, but I'm not seeing it. Curious what they do.

Reply:
we could split the component in two I guess. Or rename `useRenderProps` to something else so it isn't a hook just a helper. If we ever use a React hook in there we will need to do one of these.